ztoc/toc_builder.go

```diff
@@ -0,0 +1,207 @@
+/*
+   Copyright The Soci Snapshotter Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package ztoc
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/awslabs/soci-snapshotter/compression"
+	"github.com/klauspost/compress/zstd"
+)
+
+// TocBuilder builds the `TOC` part of a ztoc and works with different
+// compression algorithms (e.g., gzip, zstd).
+type TocBuilder struct{}
+
+// NewTocBuilder return a `TocBuilder` struct.
+func NewTocBuilder() TocBuilder {
+	return TocBuilder{}
+}
+
+// TocFromFile creates a `TOC` given a layer blob filename and the compression
+// algorithm used by the layer.
+func (tb TocBuilder) TocFromFile(algorithm, filename string) (TOC, compression.Offset, error) {
```

Comment:
How about passing an `io.Reader` here instead of `filename`?

Reply:
Do you mean only for `TocBuilder.TocFromFile` or both `TocBuilder.TocFromFile` and `ZinfoBuilder.ZinfoFromFile`?

I want to keep the index-related change minimum in this PR and only focus on ztoc/compression decouple. Currently the main change on soci index side is basically 1 line:

```golang
// within func (b *IndexBuilder) buildSociLayer

// before
toc, err := ztoc.BuildZtoc(tmpFile.Name(), b.config.spanSize, b.config.buildToolIdentifier)
// after
toc, err := b.ztocBuilder.BuildZtoc(tmpFile.Name(), b.config.spanSize, b.config.buildToolIdentifier, ztoc.WithCompression(compression))
```

I think changing `filename` to `io.Reader` itself is a big change since it comes from soci index builder down to `tocbuilder` and `zinfobuilder` (which down to zinfo compression impl, e.g., `NewGzipZinfoFromFile(gzipFile string, spanSize int64) (*GzipZinfo, error)`).

Will open a separate issue to track that.

Reply:
For now, just `TocFromFile`. `ZinfoFromFile` is a more involved change since it would require updating the C library. But the Go way is passing around `io.Readers`, not filenames.

Reply:
If only `TocFromFile` I'd prefer passing filename to both `TocFromFile` and `ZinfoFromFile` now for consistency:

1. The filename is passed from index builder to ztoc builder anyway (and used by zinfo builder).
2. Currently, ztoc builder is a thin warp of zinfo builder and toc builder (no other logic than, e.g., compression support check). This makes logic separation more clearly.
3. If change `TocFromFile` only, the difference will be just moving [`build compress io.Reader from filename` logic](https://github.com/awslabs/soci-snapshotter/blob/62e5d5bc7517ac01779df6b24a5bead8aba261bb/ztoc/toc_builder.go#L79) up to ztoc builder from toc builder.

Agree that passing `io.Reader`s is more Go way but I didn't see much benefit of only changing `Toc` while keeping `Zinfo`. Probably we should change this together starting from index builder to both toc and zinfo.

WDYT @rdpsin

Reply:
Sure. Can we have an issue to track this?

Reply:
Sure, will open an issue later with details :)